content/300-guides/300-upgrade-guides/200-upgrading-versions/030-upgrading-to-prisma-4.mdx

```diff
@@ -0,0 +1,418 @@
+```
+
+#### Upgrade path
+
+From version 4.0.0, some data types returned by `queryRaw` or `queryRawUnsafe` are different, as follows:
+
+| Data type  | Before version 4.0.0  | From version 4.0.0    |
+| ---------- | --------------------- | --------------------- |
+| `DateTime` | Returned as `String`  | Returned as `Date`    |
+| `Numeric`  | Returned as `Float`   | Returned as `Decimal` |
+| `Bytes`    | Returned as `String`  | Returned as `Buffer`  |
+| `Int64`    | Returned as `Integer` | Returned as `BigInt`  |
+
+If you use `queryRaw` or `queryRawUnsafe` to return any of the above data types, then you must change your code to handle the new types.
+
+For example, if you return `DateTime` data, then you need to take into account the following:
+
+- You no longer need to manually instantiate a `DateTime` object for the returned data.
+- If your code currently uses the returned `String` data, then you now need to convert the `DateTime` object to a `String`.
+
+You must make equivalent code changes for the other data types in the table above.
+
+#### Raw query mapping: PostgreSQL type-casts
+
+In versions 3.14.x and 3.15.x, [raw query type mapping](/concepts/components/prisma-client/raw-database-access#raw-query-type-mapping) was available with the Preview feature `improvedQueryRaw`. In version 4.0.0, we have made raw query type mapping Generally Available. You do not need to use `improvedQueryRaw` to get this functionality in versions 4.0.0 or later.
+
+Before version 4.0.0, many PostgreSQL type-casts did not work. We have tightened the type coercion rules so that all type-casts now work. As a result, some implicit casts now fail.
+
+#### Upgrade path
+
+We recommend that you re-test your use of `$queryRaw` to ensure that the types you pass into your raw queries match the types that PostgreSQL expects.
+
+For example, in version 4.0.0, the following query fails:
+
+```js
+await prisma.$queryRaw`select length(${42});
+  // ERROR: function length(integer) does not exist
+  // HINT: No function matches the given name and argument types. You might need to add explicit type casts.
+```
+
+This is because PostgreSQL’s `length` function expects `text` as input. Prisma used to silently coerce `42` to `text`, but does not do this in version 4.0.0. To fix this, explicitly cast `42`to `text` as follows:
+
+```js
+await prisma.$queryRaw`select length(${42}::text);
+```
+
+### <inlineCode>DbNull</inlineCode>, <inlineCode>JsonNull</inlineCode> and <inlineCode>AnyNull</inlineCode> are now objects
+
+JavaScript `null` is ambiguous for JSON columns, so Prisma uses `DbNull`, `JsonNull`, and `AnyNull` to distinguish between the database `NULL` value and the JSON `null` value. Before version 4.0.0, `DbNull`, `JsonNull`, and `AnyNull` were string constants. From version 4.0.0, they are objects.
+
+See [Filtering by null values](/concepts/components/prisma-client/working-with-fields/working-with-json-fields#filtering-by-null-values) for more information.
```

Comment:
```suggestion
See [filtering by null values](/concepts/components/prisma-client/working-with-fields/working-with-json-fields#filtering-by-null-values) for more information.
```

Reply:
I thought this was a mistake, but I see capitalization consistently applied, so please disregard if it's intentional.

Reply:
There's an argument for both forms, but I think we're going with capitalizing as the destination header.  Definitely a topic for our upcoming style guide discussions.